source/common/upstream/load_balancer_impl.cc

```diff
@@ -521,20 +521,33 @@ void ZoneAwareLoadBalancerBase::regenerateLocalityRoutingStructures() {
   // residual_capacity: 0 10000 15000
   // Now to find a locality to route (bucket) we could simply iterate over residual_capacity
   // searching where sampled value is placed.
-  state.residual_capacity_.resize(num_localities);
+  state.residual_capacity_.resize(num_upstream_localities);
 
-  // Local locality (index 0) does not have residual capacity as we have routed all we could.
-  state.residual_capacity_[0] = 0;
-  for (size_t i = 1; i < num_localities; ++i) {
+  auto percentage_index_it = locality_percentages->upstream_to_percentage_index.begin();
+  for (uint64_t i = 0; i < num_upstream_localities; ++i) {
+    uint64_t last_residual_capacity = i > 0 ? state.residual_capacity_[i - 1] : 0;
+    if (percentage_index_it->first != i) {
+      // Locality with index "i" does not have any upstream hosts, but we keep accumulating previous
+      // values to make search easier on the next step.
+      state.residual_capacity_[i] = last_residual_capacity;
+      continue;
+    }
+
+    uint64_t combined_locality_index = percentage_index_it->second;
     // Only route to the localities that have additional capacity.
-    if (upstream_percentage[i] > local_percentage[i]) {
+    if (locality_percentages->upstream_percentage[combined_locality_index] >
+        locality_percentages->local_percentage[combined_locality_index]) {
       state.residual_capacity_[i] =
-          state.residual_capacity_[i - 1] + upstream_percentage[i] - local_percentage[i];
+          last_residual_capacity +
+          locality_percentages->upstream_percentage[combined_locality_index] -
+          locality_percentages->local_percentage[combined_locality_index];
     } else {
       // Locality with index "i" does not have residual capacity, but we keep accumulating previous
       // values to make search easier on the next step.
-      state.residual_capacity_[i] = state.residual_capacity_[i - 1];
+      state.residual_capacity_[i] = last_residual_capacity;
     }
+
+    ++percentage_index_it;
   }
 }
```

Comment:
The result is the same as before (populating `state.residual_capacity_`), but we have to account for the slightly different format of the new percentages structure. The percentages includes a mapping that describes which combined locality percentages correspond to a given upstream locality index.